packages/toolkit/src/query/core/buildThunks.ts

```diff
@@ -255,6 +270,62 @@ export function buildThunks<
       return ret
     }
 
+  const upsertQueryData: UpsertQueryDataThunk<EndpointDefinitions, State> =
+    (endpointName, args, upsertRecipe) => (dispatch, getState) => {
+      const currentState = (
+        api.endpoints[endpointName] as ApiEndpointQuery<any, any>
+      ).select(args)(getState())
+      let ret: PatchCollection = {
+        patches: [],
+        inversePatches: [],
+        undo: () =>
+          dispatch(
+            api.util.patchQueryData(endpointName, args, ret.inversePatches)
+          ),
+      }
+      if ('data' in currentState) {
+        if (isDraftable(currentState.data)) {
+          const [, patches, inversePatches] = produceWithPatches(
+            currentState.data,
+            upsertRecipe
+          )
+          ret.patches.push(...patches)
+          ret.inversePatches.push(...inversePatches)
+        } else {
+          const value = upsertRecipe(currentState.data)
+          ret.patches.push({ op: 'replace', path: [], value })
+          ret.inversePatches.push({
+            op: 'replace',
+            path: [],
+            value: currentState.data,
+          })
+        }
+        dispatch(api.util.patchQueryData(endpointName, args, ret.patches))
+      } else {
+        ret.inversePatches.push({
```

Comment:
I'm not sure about this part tbh. - undoing an upsert would be what exactly? 
Keeping the cache entry but replacing the value with `undefined` seems wrong. Removing the cache entry could have even more implications.

=> I would suggest that this method is not in any way undoable. Just chuck a value in there and be done with it.

In the same sense, I'm not sure if we really need an "upsertRecipe". Tbh., I'd be perfectly fine with this method just "putting a value into the store, overriding what is there". 
To me it feels like it's not really the place for fine-grained, undoable updates.

I'm open to arguments though - do you think it is a good idea - do you have use cases?

Reply:
Changing the signature of `updateQueryData` is probably also not desirable because it would mean a breaking change. I think the insertMethod could be something that is used with pessimistic updates only, but I think a different name would be more appropriate in this case. Maybe something along the lines of `primeQueryCache` but that would only signal the use case of adding data to an empty cache. 

Maybe `primeQueryCache` should only do something if the query wasn't executed before. This way you could call `updateQueryData` in an optimistic fashion and if the use case is that data should also be entered if the cache is empty `primeQueryCache` can be called after the query has successfully completed. This way the user also wouldn't need to check the current query states themselves.

Reply:
Thanks for the feedback. 
I should be able to add the changes by the end of the week.

Reply:
Something came up on the weekend, will make the changes in the next couple of days